skbio/diversity/_fast_base_cy.pyx

```diff
@@ -0,0 +1,215 @@
+        ``a``, which represents a node in a tree. The second column is the 
+        starting index in ``a`` for the node's children, and the third column
+        is the ending index in ``a`` for the node's children.
+    a : np.ndarray of int
+        A matrix of the environment data. Each row corresponds to a node in a
+        tree, and each column corresponds to an environment. On input, it is
+        assumed that only tips have counts.
+
+    Notes
+    -----
+    This is effectively a postorder reduction over the tree. For example, 
+    given the following tree:
+
+                            /-A
+                  /E-------|
+                 |          \-B
+        -root----|
+                 |          /-C
+                  \F-------|
+                            \-D
+
+    And assuming counts for [A, B, C, D] in environment FOO of [1, 1, 1, 0] and
+    counts for environment BAR of [0, 1, 1, 1], the input counts matrix ``a``
+    would be:
+
+        [1 0  -> A
+         1 1  -> B
+         1 1  -> C
+         0 1  -> D
+         0 0  -> E
+         0 0  -> F
+         0 0] -> root
+
+    The method will perform the following reduction:
+
+        [1 0     [1 0     [1 0     [1 0 
+         1 1      1 1      1 1      1 1 
+         1 1      1 1      1 1      1 1 
+         0 1  ->  0 1  ->  0 1  ->  0 1 
+         0 0      2 1      2 1      2 1 
+         0 0      0 0      1 2      1 2 
+         0 0]     0 0]     0 0]     3 3]
+
+    The index positions of the above are encoded in ``child_index`` which 
+    describes the node to aggregate into, and the start and stop index 
+    positions of the nodes immediate descendents. 
+
+    This method operates inplace on ``a``
+    """
+    cdef:
+        Py_ssize_t i, j, k
+        DTYPE_t node, start, end
+        DTYPE_t n_envs = a.shape[1]
+
+    # possible GPGPU target
+    for i in range(child_index.shape[0]):
+        node = child_index[i, 0]
+        start = child_index[i, 1]
+        end = child_index[i, 2]
+       
+        for j in range(start, end + 1):
+            for k in range(n_envs):
+                a[node, k] += a[j, k]
+        
+
+@cython.boundscheck(False) 
+@cython.wraparound(False) 
+def nodes_by_counts(np.ndarray counts, 
+                    np.ndarray tip_ids, 
+                    dict indexed):
+    """Construct the count array, and the counts up the tree
+
+    Parameters
+    ----------
+    counts : np.array of int
+        A 1D or 2D vector in which each row corresponds to the observed counts
+        in an environment. The rows are expected to be in order with respect to
+        `tip_ids`.
+    tip_ids : np.array of str
+        A vector of tip names that correspond to the columns in the `counts`
+        matrix.
+    indexed : dict
+        The result of `index_tree`.
+
+    Returns
+    -------
+    np.array of int
+        The observed counts of every node and the counts if its descendents.
```

Comment:
Close, but it's in post order so `o3`, `o2`, `o1`. This documentation overlaps with the documentation in `traverse_reduce` but this comment suggests the doc is insufficient here.